dtfs-central-api/src/v1/controllers/utilisation-report-service/put-utilisation-report-status.controller/index.test.ts

```diff
@@ -113,10 +100,19 @@ describe('put-utilisation-report-status.controller', () => {
     manager: mockTransactionManager,
   } as unknown as QueryRunner;
 
+  const utilisationReportRepoFindOneByOrFailSpy = jest.spyOn(UtilisationReportRepo, 'findOneByOrFail');
+
+  const mockFindOneByOrFail = (reports: UtilisationReportEntity[]) => (where: Parameters<typeof UtilisationReportRepo.findOneByOrFail>[0]) => {
```

Comment:
Could we put this and the consts above at the top of their describe block before any tests? It's nice to keep them all in one place.

Reply:
I usually prefer to define them as they are needed as I think it makes it easier to read

Reply:
I actually find it harder to read 😅. I think it is more likely to cause confusion written this way. I rarely read a test file in order except when reviewing a freshly written test file and I would expect to see the test set up and mocking at the top of the relevant describe block rather than find it sprinkled throughout.

Reply:
I actually wonder if we want to assert in the above tests that we don't actually call the state machine and update the statuses?

Also, if the failure case test breaks and the invalid request is not caught correctly, we don't actually want to then be calling all the way through to the database. In this specific case nothing bad will happen, but I think that is at least part of the reason the convention is to put one's set up at the top.

Reply:
I think this is just a difference in opinion then as I almost always read test files in order. The compromise here I think it to wrap this part in another `describe('when the payload is valid', ...)` block which distinguishes it from the other tests, so then these mocks can be defined at the top of that block and it doesn't appear to be randomly spread out

Reply:
This isn't a difference in opinion, it is actually best practice to put these mocks at the top. In fact I've just been thinking about this more and the mocks you set up halfway down the tests are actually being set up before the test runs in most likeliness, so reading the top of this file you would have no idea the first two tests have mocks in them but they do!

If you run a test file with the following (as I have just done to make sure I am not in fact wrong 😅 ) you should get one, then three, then two printed to the console.

```
console.log('ONE');

it('fake test', () => {
  console.log('TWO');
})

console.log('BANANA THREE');
```
If you really do want to go with the multiple describe blocks (which I think is unnecessary) then please put the invalid payload ones inside an 'when the payload is invalid' describe block also.

Reply:
Yes, `jest` runs tests asynchronously (technically, concurrently).

I haven't seen this listed anywhere as a best-practice, and the concern about the later-defined values being available is mitigated somewhat by the `"no-use-before-define"` eslint rule. However, if we are going to have every single mock defined at the top, then the tests should really be using those mocks. That's not a problem in this case as we can just test the the query runner is never initialised